File: plugins/tokenmonitor/server/go/internal/textutil/textutil.go
// Package textutil holds tiny string helpers shared across the broker.
package textutil

// ClipRunes truncates s to at most n Unicode code points. The wire contract
// for 15-char fields (pet_name, quota-group labels) counts CODE POINTS —
// matching the Python implementation's s[:15] — so byte-based s[:n] (which
// can split a multi-byte UTF-8 sequence and emit invalid UTF-8 on the wire)
// must not be used for device-visible strings.
func ClipRunes(s string, n int) string {
	if n <= 0 {
		return ""
	}
	i := 0
	for pos := range s {
		if i == n {
			return s[:pos]
		}
		i++
	}
	return s
}

File: plugins/tokenmonitor/server/js/src/textutil.js
// Tiny string helpers shared across the broker.

// clipCodePoints truncates s to at most n Unicode CODE POINTS. The wire
// contract for 15-char fields (pet_name, quota-group labels) counts code
// points — matching Python's s[:15] and Go's textutil.ClipRunes — so
// String.prototype.slice (which counts UTF-16 units and can split an astral
// pair, emitting a lone surrogate on the wire) must not be used for
// device-visible strings.
export function clipCodePoints(s, n) {
  s = String(s);
  if (n <= 0) return "";
  let count = 0;
  let idx = 0; // UTF-16 unit index of the cut point
  for (const cp of s) {
    if (count === n) return s.slice(0, idx);
    idx += cp.length; // 1 for BMP, 2 for an astral surrogate pair
    count += 1;
  }
  return s;
}
